### DISCUSSION TITLE
select multiple lines so that I can paste selected text below selection and run an obsidian command on the pasted text.

### DISCUSSION THREAD

**Discussion by ksdavidc** [chhoumann/quickadd#283] [Q&A]
Probably easy, I just don't know how. I want to be able to select some text and then:

Basically I want to paste a selection below itself and then modify it. Specifically, something like:
1. copy the selected (**multiline**) text.
2. paste this text BELOW the selected text (in new paragraphs)
3. then select this (**multiline**) text
4. then run an obsidian command on this text.

I have made progress on doing 2,3, and 4. For example, I have tried just creating a macro that pastes, selecting the current line, and then runs the command. This works on single line text, but not on multiline text, which is what I need.

The transition between 1 and 2 (i.e. copying, then moving to end of selection, adding paragraphs, and then running 2) is also something I would like to be able to do.

Does anyone know how to do this? Is this possible?

**Comment by chhoumann** [comment]
Apologies this went unanswered for so long.

You're right that the built-in editor commands (Select active line, Copy, Paste) are line-scoped, so they can't extend to a multiline selection. A small user script in a Macro closes the gap: it duplicates the current selection below itself, re-selects the copy, and runs whatever command you point it at.

Save this as a user script in your vault, e.g. `scripts/duplicate-and-run.js`:

```js
module.exports = async (params) => {
	const { app, obsidian } = params;
	const view = app.workspace.getActiveViewOfType(obsidian.MarkdownView);
	const editor = view?.editor ?? app.workspace.activeEditor?.editor;
	if (!editor || !editor.somethingSelected()) {
		new obsidian.Notice("Select some text first.");
		return;
	}
	const selection = editor.getSelection();
	const to = editor.getCursor("to");

	// Insert a blank line, then the duplicated block, right after the selection.
	editor.replaceRange("\n\n" + selection, to);

	// Re-select the duplicate (it starts two lines below where the selection ended).
	const start = { line: to.line + 2, ch: 0 };
	const selLines = selection.split("\n");
	const end =
		selLines.length === 1
			? { line: start.line, ch: selection.length }
			: {
					line: start.line + selLines.length - 1,
					ch: selLines[selLines.length - 1].length,
			  };
	editor.setSelection(start, end);

	// Run whatever command you want on the re-selected block:
	app.commands.executeCommandById("editor:toggle-bold");
};
```

Then create a Macro choice, add this script as its command, and toggle the Macro's own command on (the lightning bolt) so you can bind it to a hotkey.

To target a different command, swap the id in the last line. To find a command's id, open the developer console (Ctrl/Cmd+Shift+I), type `app.commands.commands`, and look through the keys - the id is the object key (e.g. `editor:toggle-bold`, `editor:toggle-blockquote`).

Selecting `Line A` + `Line B` and running the macro left the original untouched and dropped a copy right below it with the command applied - the re-selection lands exactly on the duplicated block, so the command only affects the copy. Works best when you select whole lines.

**Reply by ksdavidc** [reply]
Thank for that. I am sad to say I have forgotten what I was originally trying to do but it may come to me as it sounds like something I would need to do. I'll let you know if it works...
